pkg/frontend/transport/handler.go

```diff
@@ -440,13 +441,31 @@ func writeError(w http.ResponseWriter, err error) {
 		}
 	}
 
-	// if the error is an APIError, ensure it gets written as a JSON response
-	if resp, ok := apierror.HTTPResponseFromError(err); ok {
-		_ = httpgrpc.WriteResponse(w, resp)
+	var (
+		res      *httpgrpc.HTTPResponse
+		resFound bool
+	)
+
+	// If the error is an APIError, ensure it gets written as a JSON response.
+	// Otherwise, check if there's a response encoded in the gRPC error.
+	res, resFound = apierror.HTTPResponseFromError(err)
+	if !resFound {
+		res, resFound = httpgrpc.HTTPResponseFromError(err)
+	}
+
+	// If we've been able to get the HTTP response from the error, then we send
+	// it with the right status code and response body content.
+	if resFound {
+		statusCode = int(res.Code)
+		_ = httpgrpc.WriteResponse(w, res)
 		return
 	}
 
-	httpgrpc.WriteError(w, err)
+	// Otherwise, we do fallback to a 5xx error, returning the non-formatted error
+	// message in the response body.
+	statusCode = http.StatusInternalServerError
```

Comment:
nit: I think it works better as a fallback if you put this line at the top of the function, then any time we cannot or miss updating it, the result will be http.StatusInternalServerError

Reply:
I see the problem but I'm not convinced that having a default is the right thing to do. I want to make sure that we always return the right status code. To solve it, I took a different approach and I removed the named return parameter so that it's intentional what we return: https://github.com/grafana/mimir/pull/8407/commits/793436b5e058c04df1938482a6b865f3eb933ef5